crates/router/src/core/payments/operations/payment_update_metadata.rs

```diff
@@ -0,0 +1,303 @@
+
+        let business_profile = db
+            .find_business_profile_by_profile_id(key_manager_state, key_store, profile_id)
+            .await
+            .to_not_found_response(errors::ApiErrorResponse::ProfileNotFound {
+                id: profile_id.get_string_repr().to_owned(),
+            })?;
+
+        let request_metadata = request.metadata.clone().expose();
+        let payment_intent_metadata = payment_intent
+            .metadata
+            .take()
+            .unwrap_or(request_metadata.clone());
+        let merged_metadata = merge_metadata(payment_intent_metadata, request_metadata);
+        payment_intent.metadata = Some(merged_metadata);
+
+        let payment_data = PaymentData {
+            flow: PhantomData,
+            payment_intent,
+            payment_attempt,
+            currency,
+            amount,
+            email: None,
+            mandate_id: None,
+            mandate_connector: None,
+            customer_acceptance: None,
+            token: None,
+            token_data: None,
+            setup_mandate: None,
+            address: payments::PaymentAddress::new(None, None, None, None),
+            confirm: None,
+            payment_method_data: None,
+            payment_method_info: None,
+            force_sync: None,
+            refunds: vec![],
+            disputes: vec![],
+            attempts: None,
+            sessions_token: vec![],
+            card_cvc: None,
+            creds_identifier: None,
+            pm_token: None,
+            connector_customer_id: None,
+            recurring_mandate_payment_data: None,
+            ephemeral_key: None,
+            multiple_capture_data: None,
+            redirect_response: None,
+            surcharge_details: None,
+            frm_message: None,
+            payment_link_data: None,
+            incremental_authorization_details: None,
+            authorizations: vec![],
+            authentication: None,
+            recurring_details: None,
+            poll_config: None,
+            tax_data: None,
+            session_id: None,
+            service_details: None,
+            card_testing_guard_data: None,
+            vault_operation: None,
+            threeds_method_comp_ind: None,
+        };
+        let get_trackers_response = operations::GetTrackerResponse {
+            operation: Box::new(self),
+            customer_details: None,
+            payment_data,
+            business_profile,
+            mandate_type: None,
+        };
+
+        Ok(get_trackers_response)
+    }
+}
+
+fn merge_metadata(
+    mut payment_intent_metadata: serde_json::Value,
+    request_metadata: serde_json::Value,
+) -> serde_json::Value {
+    if let (serde_json::Value::Object(existing_map), serde_json::Value::Object(req_map)) =
+        (&mut payment_intent_metadata, request_metadata.clone())
+    {
+        for (key, value) in req_map {
+            existing_map.insert(key, value);
+        }
```

Comment:
Can't we use `extend`?

```suggestion
        existing_map.extend(req_map);
```

As per the [trait docs](https://doc.rust-lang.org/core/iter/trait.Extend.html):

> When extending a collection with an already existing key, that entry is updated or, in the case of collections that permit multiple entries with equal keys, that entry is inserted.

So this should overwrite existing keys (same behavior as existing `insert` based implementation).